fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/metadata/RecordMetaDataBuilderTest.java

```diff
@@ -224,4 +227,58 @@ public void badUnionFields() {
+
+        // Make a local version of the meta-data dynamically
+        RecordMetaDataProto.MetaData.Builder serializedLocalMetaDataBuilder = RecordMetaDataProto.MetaData.newBuilder()
+                .setRecords(originalMetaData.getRecords());
+        serializedLocalMetaDataBuilder.addDependencies(TestRecords2Proto.getDescriptor().toProto());
+        serializedLocalMetaDataBuilder.addRecordTypes(serializedLocalMetaDataBuilder
+                                                            .addRecordTypesBuilder()
+                                                            .setName("MyLongRecord").build());
+        serializedLocalMetaDataBuilder.setRecords(serializedLocalMetaDataBuilder
+                                                            .getRecordsBuilder()
+                                                            .addMessageType(TestRecords2Proto.MyLongRecord.getDescriptor().toProto())
+                                                            .build());
+        RecordMetaDataProto.MetaData serializedLocalMetaData = serializedLocalMetaDataBuilder.build();
+        assertTrue(serializedLocalMetaData.getRecordTypesList().stream()
+                .map(RecordMetaDataProto.RecordType::getName)
+                .filter("MyLongRecord"::equals).findAny().isPresent());
+        RecordMetaDataBuilder localMetaDataBuilder = RecordMetaData.newBuilder()
+                .addDependency(TestRecords2Proto.getDescriptor())
+                .setRecords(serializedLocalMetaData);
+        assertNotNull(localMetaDataBuilder.getRecordType("MyLongRecord"));
+        RecordMetaData localMetaData = localMetaDataBuilder.build();
+
+        // Create a new builder of the original meta-data
+        RecordMetaDataBuilder newMetaDataBuilder = RecordMetaData.newBuilder()
+                .setLocalFileDescriptor(localMetaData.getRecordsDescriptor())
+                .setRecords(originalMetaData);
+        Throwable t = assertThrows(MetaDataException.class, () ->
+                newMetaDataBuilder.getRecordType("MyLongRecord"));
+        assertEquals("Unknown record type MyLongRecord", t.getMessage());
+        RecordMetaData newMetaData = newMetaDataBuilder.build(true);
+        assertEquals(newMetaData.toProto(null), originalMetaData);
+        assertEquals(newMetaData.getUnionDescriptor(), localMetaData.getUnionDescriptor());
+
+        // Record type moved from being imported to being present in the local descriptor
+        RecordMetaDataProto.MetaData previouslyImportedMetaData = RecordMetaData.build(TestRecordsImportProto.getDescriptor()).toProto(null);
+        RecordMetaDataBuilder nowFlatMetaData = RecordMetaData.newBuilder()
+                .setLocalFileDescriptor(TestRecordsImportFlatProto.getDescriptor())
+                .setRecords(previouslyImportedMetaData);
+        assertNotNull(nowFlatMetaData.getRecordType("MySimpleRecord"));
+        assertNotNull(nowFlatMetaData.getRecordType("MyLongRecord"));
```

Comment:
It should also verify that the file descriptor of the record type is the expected descriptor.